lib/features/support/interactive_cli.rb

```diff
@@ -0,0 +1,117 @@
+require 'open3'
+
+# Encapsulates a terminal session, retaining state and input streams for interactive tests
+class InteractiveCLI
+  # @!attribute [r] parsed_output
+  #   @return [Array] An array of output strings received from the terminals STDOUT pipe
+  attr_reader :parsed_output
+
+  # @!attribute [r] parsed_error
+  #   @return [Array] An array of error strings received from the terminals STDERR pipe
+  attr_reader :parsed_errors
+
+  # @!attribute [r] last_exit_code
+  #   @return [Number] The exit code of the last terminal command
+  attr_reader :last_exit_code
+
+  # @!attribute [r] pid
+  #   @return [Number] The PID of the running terminal
+  attr_reader :pid
+
+  # @!attribute [r] current_buffer
+  #   @return [String] A string representation of the current output present in the terminal
+  attr_reader :current_buffer
+
+  # Creates an InteractiveCLI instance
+  #
+  # @param environment [Hash] A hash of environment variables
+  # @param terminal [String] A path to the terminal to run, defaults to `/bin/sh`
+  # @param terminal [String] The stop command, defaults to `exit`
+  def initialize(environment = {}, terminal = '/bin/sh', stop_command = 'exit')
+    @env = environment
+    @parsed_output = []
+    @parsed_errors = []
+    @current_buffer = ''
+    @last_exit_code = nil
+    @in_stream = nil
+    @pid = nil
+    @running = false
+    @stop_command = stop_command
+    start_threaded_terminal(terminal)
+  end
+
+  # Attempts to stop the terminal using the preset command
+  #
+  # @return [Boolean] true if successful, false otherwise
+  def stop
+    run_command(@stop_command)
+  end
+
+  # Returns whether the terminal is running by verifying the in_stream exists and is open
+  #
+  # @return [Boolean] Whether the stream is currently running
+  def running
+    !(@in_stream.nil? || @in_stream.closed?)
+  end
+
+  # Runs the given command if the terminal is running
+  #
+  # @param command [String] The command to run
+  #
+  # @return [Boolean] true if successful, false otherwise
+  def run_command(command)
+    return false unless running
+    @in_stream.puts(command)
+    true
+  end
+
+  private
+
+  # Starts a terminal on another thread
+  #
+  # @param terminal [String] A path to the terminal to run
+  def start_threaded_terminal(terminal)
```

Comment:
Does this not work properly without being run in a thread? I'm wondering if we need this method, as callers could start it in a new thread if they needed to

If it always has to be run in a thread, is it worth inlining `start_terminal` here so that it doesn't look like there are two ways to start it?

Reply:
This is one of the requirements for unit testing the class, as mocha will not function on objects running in other threads.